src/vsr/replica_format.zig

```diff
@@ -26,195 +22,287 @@ pub fn format(
-
-                for (std.mem.bytesAsSlice(Header.Prepare, wal_buffer[0..size])) |*header| {
-                    assert(header.valid_checksum());
-
-                    if (header.op == 0) {
-                        assert(header.operation == .root);
-                    } else {
-                        assert(header.operation == .reserved);
-                    }
-                }
 
-                storage.write_sectors(
-                    write_sectors_callback,
-                    &self.write,
-                    wal_buffer[0..size],
-                    .wal_headers,
-                    wal_offset,
+            // Direct I/O requires the buffer to be sector-aligned. Unlike the Prepares above that
+            // require a buffer per prepare, since they are spread out with zeros inbetween, the
+            // headers zone is contiguous so a single buffer will do.
+            const headers_buffer = try arena.alignedAlloc(
+                u8,
+                constants.sector_size,
+                vsr.sector_ceil(constants.journal_size_headers),
+            );
+            @memset(headers_buffer, 0);
+
+            for (0..constants.journal_slot_count) |slot| {
+                const single_header_buffer =
+                    headers_buffer[slot * @sizeOf(Header.Prepare) ..][0..@sizeOf(Header.Prepare)];
+                const header, const prepare_offset = wal_header(
+                    cluster,
+                    slot,
+                    single_header_buffer,
                 );
-                self.formatting = true;
-                while (self.formatting) storage.run();
-                wal_offset += size;
+                _ = prepare_offset;
+
+                assert(!stdx.zeroed(single_header_buffer));
+                assert(stdx.zeroed(headers_buffer[(slot + 1) * @sizeOf(Header.Prepare) ..]));
+                assert(header.valid_checksum());
+
+                if (header.op == 0) {
+                    assert(header.operation == .root);
+                } else {
+                    assert(header.operation == .reserved);
+                }
             }
-            // There are no headers left to write.
-            assert(format_wal_headers(cluster, wal_offset, wal_buffer) == 0);
+
+            storage.write_sectors(
+                write_sectors_callback,
+                &self.writes[self.writes_pending],
+                headers_buffer,
+                .wal_headers,
+                0,
+            );
+            self.writes_pending += 1;
         }
 
-        fn format_replies(
+        fn queue_format_replies(
             self: *ReplicaFormat,
-            allocator: std.mem.Allocator,
+            arena: std.mem.Allocator,
             storage: *Storage,
         ) !void {
-            assert(!self.formatting);
+            assert(!self.formatting and !self.formatting_superblock);
 
             // Direct I/O requires the buffer to be sector-aligned.
             const message_buffer =
```

Comment:
remove \n after `=`, to keep formatting consistent. And no blank line before memset